src/library/scala/collection/StrictOptimizedSeqOps.scala

```diff
@@ -12,17 +12,19 @@ trait StrictOptimizedSeqOps [+A, +CC[_], +C]
     with StrictOptimizedIterableOps[A, CC, C] {
 
   override def distinctBy[B](f: A => B): C = {
-    val builder = newSpecificBuilder()
-    val seen = mutable.HashSet.empty[B]
-
-    for (x <- this) {
-      val y = f(x)
-      if (!seen.contains(y)) {
-        seen += y
-        builder += x
+    val isImmutable = this.isInstanceOf[immutable.Seq[_]]
```

Comment:
This implementation should be split up into `collection.StrictOptimizedSeqOps` and `collection.immutable.StrictOptimizedSeqOps`. The only reason why we needed this flag in 2.12.x is binary compatibility.